runtime/major_gc.c

```diff
@@ -562,10 +581,176 @@ static void mark_ephe_aux (struct mark_stack *stk, intnat *work,
+#ifndef CAML_INSTR
+  (intnat work)
+#else
+  (intnat work, int* pslice_fields, int* pslice_pointers)
+#endif
+{
+  uintnat pb_enqueued = 0, pb_dequeued = 0;
+  int darkened_anything = 0;
+  value pb[Pb_size];
+  /* These global values are cached in locals,
+     so that they can be stored in registers */
+  struct mark_stack stk = *Caml_state->mark_stack;
+  uintnat young_start = (uintnat)Val_hp(Caml_state->young_start);
+  uintnat half_young_len =
+    ((uintnat)Caml_state->young_end - (uintnat)Caml_state->young_start) >> 1;
+#define Is_block_and_not_young(v) \
+  (((intnat)rotate1((uintnat)v - young_start)) >= (intnat)half_young_len)
+#ifdef NO_NAKED_POINTERS
+  #define Is_major_block(v) Is_block_and_not_young(v)
+#else
+  #define Is_major_block(v) (Is_block_and_not_young(v) && Is_in_heap(v))
+#endif
+
+#ifdef CAML_INSTR
+  int slice_fields = 0, slice_pointers = 0;
+#endif
+
+  while (1) {
+    value *scan, *obj_end, *scan_end;
+    uintnat min_pb = Pb_min; /* keep pb at least this full */
+
+    if (work <= 0 || stk.count == 0) {
+      if (pb_enqueued == pb_dequeued)
+        break;
+      /* Dequeue from pb even when close to empty, because
+         we have nothing else to do */
+      min_pb = 0;
+    }
+
+    if (pb_enqueued > pb_dequeued + min_pb) {
+      /* Dequeue from prefetch buffer */
+      value block = pb[(pb_dequeued++) & Pb_mask];
+      header_t hd = Hd_val(block);
+
+      if (Tag_hd(hd) == Infix_tag) {
+        block -= Infix_offset_val(block);
+        hd = Hd_val(block);
+      }
+
+#ifdef NO_NAKED_POINTERS
+      /* See [caml_darken] for a description of this assertion. */
+      CAMLassert (Is_in_heap (block) || Is_black_hd (hd));
+#endif
+      CAMLassert(Is_white_hd(hd) || Is_black_hd(hd));
+      if (!Is_white_hd (hd)) {
+        /* Already black, nothing to do */
+        continue;
+      }
+      hd = Blackhd_hd (hd);
+      Hd_val (block) = hd;
+      darkened_anything = 1;
+      work--; /* header word */
+      if (Tag_hd (hd) >= No_scan_tag) {
+        /* Nothing to scan here */
+        work -= Wosize_hd (hd);
+        continue;
+      }
+      scan = Op_val(block);
+      obj_end = scan + Wosize_hd(hd);
+
+      if (Tag_hd (hd) == Closure_tag) {
+        uintnat env_offset = Start_env_closinfo(Closinfo_val(block));
+        work -= env_offset;
+        scan += env_offset;
+      }
+    } else {
+      mark_entry m = stk.stack[--stk.count];
+      scan = m.start;
+      obj_end = m.end;
+    }
+
+    scan_end = obj_end;
+    work -= obj_end - scan;
+    if (work < 0) {
+      scan_end += work;
+    }
+
+    for (; scan < scan_end; scan++) {
+      value v = *scan;
+#ifdef CAML_INSTR
+      slice_fields ++;
+#endif
+      if (Is_major_block(v)) {
+#ifdef CAML_INSTR
+        slice_pointers ++;
+#endif
+        if (pb_enqueued == pb_dequeued + Pb_size) {
+          break; /* Prefetch buffer is full */
+        }
+        prefetch_block(v);
+        pb[(pb_enqueued++) & Pb_mask] = v;
+      }
+    }
```

Comment:
You've inadvertently hobbled the naked pointers checker. To restore it, you'll have to add something like this:

```suggestion
    }
#if defined(NAKED_POINTERS_CHECKER) && defined (NATIVE_CODE)
      else if (Is_block_and_not_young (v) && !Is_in_heap (v)){
        is_naked_pointer_safe (v, scan);
      }
#endif
```

Reply:
Thanks, applied.